File: JavaScript/src/romanNumerals.js
// https://www.codewars.com/kata/52a89c2ea8ddc5547a000863/javascript

class RomanNumber
{
	/** @type string */
	letter;
	
	/** @type number */
	number;
	
	/** @param {string} letter
	 * @param {number} number */
	constructor(letter, number)
	{
		this.letter = letter;
		this.number = number;
	}
}

export class RomanNumerals
{
	static #romanNumbers = [
		new RomanNumber("M", 1000),
		new RomanNumber("CM", 900),
		new RomanNumber("D", 500),
		new RomanNumber("CD", 400),
		new RomanNumber("C", 100),
		new RomanNumber("XC", 90),
		new RomanNumber("L", 50),
		new RomanNumber("XL", 40),
		new RomanNumber("X", 10),
		new RomanNumber("IX", 9),
		new RomanNumber("V", 5),
		new RomanNumber("IV", 4),
		new RomanNumber("I", 1),
	];
	
	/** @param {string} roman */
	static fromRoman(roman)
	{
		let ret = 0;
		let i = 0;
		while (roman.length)
		{
			if (roman.startsWith(this.#romanNumbers[i].letter))
			{
				ret += this.#romanNumbers[i].number;
				roman = roman.substring(this.#romanNumbers[i].letter.length);
			}
			else
			{
				i++;
			}
		}
		return ret;
	}
	
	/** @param {number} number */
	static toRoman(number)
	{
		let i = 0;
		let ret = "";
		while (number)
		{
			if (number >= this.#romanNumbers[i].number)
			{
				ret += this.#romanNumbers[i].letter;
				number -= this.#romanNumbers[i].number;
			}
			else
			{
				i++;
			}
		}
		return ret;
	}
}

File: C#/src/RomanNumerals.cs
// https://www.codewars.com/kata/51b66044bce5799a7f000003/csharp

namespace CSharp;

public record struct RomanNumber(string Letter, int Number);

public static class RomanNumerals
{
	private static readonly IList<RomanNumber> RomanNumbers = new List<RomanNumber> {
			new("M" , 1000),
			new("CM",  900),
			new("D" ,  500),
			new("CD",  400),
			new("C" ,  100),
			new("XC",   90),
			new("L" ,   50),
			new("XL",   40),
			new("X" ,   10),
			new("IX",    9),
			new("V" ,    5),
			new("IV",    4),
			new("I" ,    1),
	};

	public static string ToRoman(int n)
	{
		var i = 0;
		var ret = "";
		while (n > 0)
		{
			if (n >= RomanNumbers[i].Number)
			{
				ret += RomanNumbers[i].Letter;
				n -= RomanNumbers[i].Number;
			}
			else
			{
				i++;
			}
		}
		return ret;
	}

	public static int FromRoman(string romanNumeral)
	{
		var ret = 0;
		var i = 0;
		while (romanNumeral.Length > 0)
		{
			if (romanNumeral.StartsWith(RomanNumbers[i].Letter))
			{
				ret += RomanNumbers[i].Number;
				romanNumeral = romanNumeral[RomanNumbers[i].Letter.Length..];
			}
			else
			{
				i++;
			}
		}
		return ret;
	}
}
